lib/spack/spack/build_environment.py

```diff
@@ -813,158 +801,242 @@ def setup_package(pkg, dirty, context="build"):
+                if dt.LINK & deptypes:
+                    modes[key] |= BUILDTIME
+
+            elif context == Context.TEST:
+                if (dt.RUN | dt.TEST) & deptypes:
+                    modes[key] |= RUNTIME_EXECUTABLE
+                elif dt.LINK & deptypes:
+                    modes[key] |= RUNTIME
+
+            elif context == Context.RUN:
+                if dt.RUN & deptypes:
+                    modes[key] |= RUNTIME_EXECUTABLE
+                elif dt.LINK & deptypes:
+                    modes[key] |= RUNTIME
+
+        # Propagate RUNTIME and RUNTIME_EXECUTABLE through link and run deps.
+        if (RUNTIME | RUNTIME_EXECUTABLE | BUILDTIME_DIRECT) & mode:
+            if dt.LINK & deptypes:
+                modes[key] |= RUNTIME
+            if dt.RUN & deptypes:
+                modes[key] |= RUNTIME_EXECUTABLE
+
+        # Propagate BUILDTIME through link deps.
+        if BUILDTIME & mode:
+            if dt.LINK & deptypes:
+                modes[key] |= BUILDTIME
+
+        # Finalize the spec; the invariant is that all in-edges are processed
+        # before out-edges, meaning that edge.parent is done.
+        if not (ADDED & mode):
+            modes[edge.parent] |= ADDED
+            nodes_with_type.append((edge.parent, mode))
+
+    # Attach the leaf nodes, since we only added nodes
+    # with out-edges. Can this be improved?
+    for spec, mode in modes.items():
+        if mode and not (ADDED & mode):
+            nodes_with_type.append((spec, mode))
+
+    return nodes_with_type
+
+
+class SetupContext:
+    """This class encapsulates the logic to determine environment modifications, and is used as
+    well to set magic globals in modules of package.py."""
+
+    def __init__(self, *specs: spack.spec.Spec, context: Context) -> None:
+        """Construct a ModificationsFromDag object.
+        Args:
+            specs: single root spec for build/test context, possibly more for run context
+            context: build, run, or test"""
+        if (context == Context.BUILD or context == Context.TEST) and not len(specs) == 1:
+            raise ValueError("Cannot setup build environment for multiple specs")
+        specs_with_type = effective_deptypes(*specs, context=context)
+
+        self.specs = specs
+        self.context = context
+        self.external: List[spack.spec.Spec]
+        self.nonexternal: List[spack.spec.Spec]
+        # Reverse so we go from leaf to root
+        self.nodes_in_subdag = set(id(s) for s, _ in specs_with_type)
+
+        # Split into non-external and extenal, maintaining topo order per group.
+        self.external, self.nonexternal = stable_partition(
+            reversed(specs_with_type), lambda t: t[0].external
         )
+        self.should_be_runnable = BUILDTIME_DIRECT | RUNTIME_EXECUTABLE
+        self.should_setup_dependent_build_env = BUILDTIME | BUILDTIME_DIRECT
+        self.should_setup_run_env = RUNTIME | RUNTIME_EXECUTABLE
```

Comment:
I wonder if we can simplify this further... These values depend only on `context` :thinking:

Reply:
in principle context could be static & part of the function name. setup_run_context(...), setup_build_context(...) etc. But that causes duplication